.github/scripts/format_dbt_test_failures.py

```diff
@@ -0,0 +1,256 @@
+#!/usr/bin/env python3
+#
+# Generates an Excel workbook of dbt test failures that can be shared with
+# other teams for review and correction.
+#
+# This script assumes that it is being run in sequence after a call to
+# `dbt test --store-failures`, since it depends on two files created by
+# that operation (target/run_results.json and target/manifest.json).
+# It also requires Python dependencies be installed from
+# format_dbt_test_failures_requirements.txt.
+#
+# Accepts three optional positional arguments:
+#
+#   1. The local path to a run_results.json file generated by a test run
+#     * If not present, defaults to './target/run_results.json'
+#   2. The local path to a manifest.json file with the compiled dbt DAG
+#     * If not present, defaults to './target/manifest.json'
+#   3. The output filepath to which the workbook should be written
+#     * If not present, defaults to './qc_test_failures_<date>.xlsx'
+#
+# Expects one optional environment variable to be set:
+#
+#  1.AWS_ATHENA_S3_STAGING_DIR: Location in S3 where Athena query results
+#    should be written. If missing, defaults to the
+#    ccao-athena-results-us-east-1 bucket
+#
+# Each sheet in the output workbook represents a category of test, e.g.
+# "valid_range" or "not_null"; each row in a sheet represents a row in a
+# database that failed a test, with enough metadata that a reader can
+# figure out what conditions caused the test to fail and investigate the
+# root cause.
+#
+# Example usage:
+#
+#   AWS_ATHENA_S3_STAGING_DIR=s3://foo-bar-baz/ \
+#       python3 format_dbt_test_failures.py \
+#       ./target/run_results.json \
+#       ./target/manifest.json \
+#       ./qc_test_failures.xlsx
+
+import dataclasses
+import datetime
+import json
+import os
+import sys
+import typing
+
+import openpyxl
+import openpyxl.styles
+import pyathena
+import pyathena.cursor
+from openpyxl.worksheet.worksheet import Worksheet
+
+DEFAULT_TEST_CATEGORY = "miscellaneous"
+AWS_ATHENA_S3_STAGING_DIR = os.getenv(
+    "AWS_ATHENA_S3_STAGING_DIR", "s3://ccao-athena-results-us-east-1/"
+)
+
+
+@dataclasses.dataclass
+class FailedTestGroup:
+    """Class to store query results for a group of failed dbt tests and provide
+    convenience methods for formatting those results for output to a report."""
+
+    def __init__(self) -> None:
+        self._rows: typing.List[typing.Dict] = []
+
+    def update(self, new_rows: typing.List[typing.Dict]) -> None:
+        """Add a list of new_rows to the rows of failing tests tracked by
+        this group. The new_rows list should be formatted like the rows
+        returned by a csv.DictReader or a DictCursor, i.e. a list of
+        dicts mapping `{column_name: row_value}`."""
+        self._rows = [*self._rows, *new_rows]
+
+    @property
+    def fieldnames(self) -> typing.List[str]:
```

Comment:
**issue (maybe):** One potential hiccup here is that since the field names are only populated from failing tests, the actual columns present in the workbook could change a _lot_ between runs. I'm worried that might be kind of confusing for an end user.

One possibility is to pull the metadata from every table in a given category then use it to create the columns, regardless of whether the test failed or not. Not tied to this idea and curious to hear your thoughts!

Reply:
That sounds good to me!